src/output_writers.jl

```diff
@@ -270,126 +270,217 @@ end
-                            atts=u_attr, compress=fw.compression)
+    # Initiates the output file with dimensions
+    write_grid(model; filename=filename, mode=mode,
+               compression=compression, attrib=globalattrib, slice_kw...)
 
-    nccreate(filepath, "v", "xC", xC, xC_attr,
-                            "yF", yC, yC_attr,
-                            "zC", zC, zC_attr,
-                            atts=v_attr, compress=fw.compression)
+    # Opens the same output file for writing fields from the user-supplied variable outputs
+    dataset = Dataset(filename, "a")
 
-    nccreate(filepath, "w", "xC", xC, xC_attr,
-                            "yC", yC, yC_attr,
-                            "zF", zC, zC_attr,
-                            atts=w_attr, compress=fw.compression)
+    # Creates an unliimited dimension "Time"
+    defDim(dataset, "Time", Inf); sync(dataset)
+    defVar(dataset, "Time", Float64, ("Time",)); sync(dataset)
+    nt = 0 # Number of time-steps
 
-    nccreate(filepath, "T", "xC", xC, xC_attr,
-                            "yC", yC, yC_attr,
-                            "zC", zC, zC_attr,
-                            atts=T_attr, compress=fw.compression)
+    if isa(outputattrib, Nothing)
+        outputattrib = Dict("u" => Dict("longname" => "Velocity in the x-direction", "units" => "m/s"),
+                            "v" => Dict("longname" => "Velocity in the y-direction", "units" => "m/s"),
+                            "w" => Dict("longname" => "Velocity in the z-direction", "units" => "m/s"),
+                            "T" => Dict("longname" => "Temperature", "units" => "K"),
+                            "S" => Dict("longname" => "Salinity", "units" => "g/kg"))
+    end
 
-    nccreate(filepath, "S", "xC", xC, xC_attr,
-                            "yC", yC, yC_attr,
-                            "zC", zC, zC_attr,
-                            atts=S_attr, compress=fw.compression)
+    # Initiates empty Float32 arrays for fields from the user-supplied variable outputs
+    for (fieldname, field) in outputs
+        dtype = eltype(parentdata(field))
+        defVar(dataset, fieldname, dtype, (netcdf_spatial_dimensions(field)...,"Time"),
+               compression=compression, attrib=outputattrib[fieldname])
+    end
+    sync(dataset)
 
-    ncwrite(u, filepath, "u")
-    ncwrite(v, filepath, "v")
-    ncwrite(w, filepath, "w")
-    ncwrite(T, filepath, "T")
-    ncwrite(S, filepath, "S")
+    # Stores slices for the dimensions of each output field
+    slices = Dict{String, Vector{Union{OrdinalRange,Colon}}}()
+    for (fieldname, field) in outputs
+        slices[fieldname] = slice(field; slice_kw...)
+    end
 
-    ncclose(filepath)
+    return NetCDFOutputWriter(filename, dataset, outputs, interval,
+                          frequency, clobber, slices, nt, 0.0)
+end
 
-    return
+
+# Closes the outputwriter
+function OWClose(fw::NetCDFOutputWriter)
+    close(fw.dataset)
 end
```

Comment:
Hmmm, `OWClose` feels a little obscure and I think violates the Julia convention that functions should be lower case. Any reason we can't dispatch like
```julia
close(fw::NetCDFOutputWriter) = close(fw.dataset)
```

Reply:
What is the function that opens the file, and why is the file not closed within this function?

Reply:
Fixed. I had to overload `Base.close`.

Reply:
> What is the function that opens the file, and why is the file not closed within this function?

Sorry, missed this question. It is opened when `NetCDFOutputWriter` is created ([here](https://github.com/climate-machine/Oceananigans.jl/blob/outputwriters_with_assoc_dimensions/src/output_writers.jl#L411)). Opening and closing netcdf files is expensive, so we avoid opening and closing it at every write iteration.